source/MaterialXGenShader/ShaderStage.h

```diff
@@ -0,0 +1,303 @@
+
+    /// Return a variable by name. Returns nullptr if
+    /// no variable is found by the given name.
+    const ShaderPort* find(const string& name) const;
+
+    /// Add a new shader port to this block.
+    ShaderPort* add(const TypeDesc* type, const string& name, ValuePtr value = nullptr);
+
+    /// Add an existing shader port to this block.
+    void add(ShaderPortPtr port);
+
+  private:
+    string _name;
+    string _instance;
+    std::unordered_map<string, ShaderPortPtr> _variableMap;
+    vector<ShaderPort*> _variableOrder;
+};
+
+
+/// A shader stage, containing the state and 
+/// resulting source code for the stage.
+class ShaderStage
+{
+  public:
+    /// Bracket types
+    enum class Brackets
+    {
+        NONE,
+        BRACES,
+        PARENTHESES,
+        SQUARES
+    };
+
+    /// Contructor.
+    ShaderStage(const string& name, ConstSyntaxPtr syntax);
+
+    /// Return the stage name.
+    const string& getName() const { return _name; }
+
+    /// Return the stage source code.
+    const string& getSourceCode() const { return _code; }
+
+    /// Create a new uniform variable block.
+    VariableBlockPtr createUniformBlock(const string& name, const string& instance = EMPTY_STRING);
+
+    /// Create a new input variable block.
+    VariableBlockPtr createInputBlock(const string& name, const string& instance = EMPTY_STRING);
+
+    /// Create a new output variable block.
+    VariableBlockPtr createOutputBlock(const string& name, const string& instance = EMPTY_STRING);
+
+    /// Return the uniform variable block with given name.
+    VariableBlock& getUniformBlock(const string& name);
+
+    /// Return the uniform variable block with given name.
+    const VariableBlock& getUniformBlock(const string& name) const;
+
+    /// Return the input variable block with given name.
+    VariableBlock& getInputBlock(const string& name);
+
+    /// Return the input variable block with given name.
+    const VariableBlock& getInputBlock(const string& name) const;
+
+    /// Return the output variable block with given name.
+    VariableBlock& getOutputBlock(const string& name);
+
+    /// Return the output variable block with given name.
+    const VariableBlock& getOutputBlock(const string& name) const;
+
+    /// Return the constant variable block.
+    VariableBlock& getConstantBlock();
+
+    /// Return the constant variable block.
+    const VariableBlock& getConstantBlock() const;
+
+    /// Return a map of all uniform blocks.
+    const VariableBlockMap& getUniformBlocks() const
+    {
+        return _uniforms;
+    }
+
+    /// Return a map of all input blocks.
+    const VariableBlockMap& getInputBlocks() const
+    {
+        return _inputs;
+    }
+
+    /// Return a map of all output blocks.
+    const VariableBlockMap& getOutputBlocks() const
+    {
+        return _outputs;
+    }
+ 
+protected:
+    /// Start a new scope using the given bracket type.
+    void beginScope(Brackets brackets = Brackets::BRACES);
```

Comment:
Shader stage source code editing is protected and only accessible through the ShaderGenerator,